Fix GCC 5 Builds

ax3l opened the pull request:
<!--Thank you so much for your PR! To help us review, fill out the form
to the best of your ability.  Please make use of the development guide at
http://yt-project.org/docs/dev/developing/index.html-->

<!--Provide a general summary of your changes in the title above, for
example "Raises ValueError on Non-Numeric Input to set_xlim".  Please avoid
non-descriptive titles such as "Addresses issue #8576".-->

<!--If you are able to do so, please do not create the PR out of master, but out
of a separate branch. -->

## PR Summary

Use GCC 5 to trigger compiler errors with non-default  C++14 compilers.

Follow-up to regressions from #2575 and #2610
Fix #2892

```
export CC=$(which gcc-5)
export CXX=$(which g++-5)
python3 -m pip wheel -v .
```

## PR Checklist

<!-- Note that some of these check boxes may not apply to all pull requests -->

- [ ] pass `black --check yt/`
- [ ] pass `isort . --check --diff`
- [ ] pass `flake8 yt/`
- [ ] pass `flynt yt/ --fail-on-change --dry-run -e yt/extern`
- [ ] New features are documented, with docstrings and narrative docs
- [ ] Adds a test for any bugs fixed. Adds tests for new features.

<!--We understand that PRs can sometimes be overwhelming, especially as the
reviews start coming in.  Please let us know if the reviews are unclear or the
recommended next step seems overly demanding , or if you would like help in
addressing a reviewer's comments.  And please ping us if you've been waiting
too long to hear back on your PR.-->

ax3l commented on .travis.yml:
```diff
@@ -17,11 +17,14 @@ addons:
   global:
     PIP=pip
     HOMEBREW_NO_AUTO_UPDATE=1
+    CC=gcc-5
```
changes in this file are temporary, unless we want to add coverage in CI for older compilers permanently

ax3l commented:
Hi @matthewturk @cphyc,

I saw your PRs in #2575 and #2610 and am not an expert in the places I touch here. Help & hints welcome :)

ax3l commented:
@RevathiJambunathan does this PR work for you, with respect to #2892?
```bash
python3 -m pip uninstall yt
pip install git+git://github.com/ax3l/yt.git@fix-gcc5
```

ax3l commented on tests/travis_install.sh:
```diff
@@ -42,4 +43,4 @@ fi
 # Step 3: install yt
 $PIP install -e .
```
fyi: there is no editable install with PEP517 projects anymore [[1]](https://www.python.org/dev/peps/pep-0517/#get-requires-for-build-sdist) [[2]](https://github.com/pypa/pip/issues/7953#issuecomment-646588548)

cphyc replied:
Oh, I didn't know that! I believe a simple fix would be to use `$PIP install .` instead, right?

ax3l replied:
Just found out as well.

According to my understanding, PEP517 packages should be installed locally like this:
```bash
python3 -m pip wheel .
python3 -m pip install *whl
```

brave new world, I guess.

neutrinoceros replied:
wow. Sorry for asking the naive question, but by "PEP517 project" you mean any project with a `pyproject.toml` file, right ?

ax3l replied:
Yep... I know...

neutrinoceros replied:
thanks for bringing this up, I was the person who introduced pyproject.toml to the project but wasn't aware of this :/

cphyc commented:
It seems the bug triggered on OSX may be fixed by replacing `-fopenmp` by `-fopenmp=libomp` (https://stackoverflow.com/questions/47822547/non-default-compilers-gcc-or-vanilla-clang-on-macos-high-sierra-linking-issue/50313990#50313990). I don't understand why it is failing now and used to work before?

cphyc commented:
I am sorry, I pushed onto your branch by mistake. I meant to push on my fork instead…

ax3l commented on setup.py:
```diff
@@ -36,8 +37,20 @@
     long_description = file.read()
 
+OMP_CONFIG = defaultdict(
+    lambda: ["--fopenmp"], {"unix": ["--fopenmp"], "msvc": ["/openmp"]}
```
```suggestion
    lambda: ["-fopenmp"], {"unix": ["-fopenmp"], "msvc": ["/openmp"]}
```

ax3l commented on setup.py:
```diff
@@ -36,8 +37,20 @@
+    lambda: ["--fopenmp"], {"unix": ["--fopenmp"], "msvc": ["/openmp"]}
+)
+CPP14_CONFIG = defaultdict(
+    lambda: ["--std=c++14"], {"unix": ["--std=c++14"], "msvc": ["/std:c++14"]}
```
```suggestion
    lambda: ["-std=c++14"], {"unix": ["-std=c++14"], "msvc": ["/std:c++14"]}
```

ax3l commented on setup.py:
```diff
@@ -36,8 +37,20 @@
+    lambda: ["--std=c++14"], {"unix": ["--std=c++14"], "msvc": ["/std:c++14"]}
+)
+CPP03_CONFIG = defaultdict(
+    lambda: ["--std=c++03"], {"unix": ["--std=c++03"], "msvc": ["/std:c++03"]}
```
```suggestion
    lambda: ["-std=c++03"], {"unix": ["-std=c++03"], "msvc": ["/std:c++03"]}
```

ax3l reviewed (commented):
The compiler flags are wrong.

ax3l reviewed (commented):
Fix compiler flags for OpenMP and C++ std.

ax3l commented:
@cphyc can you please add the `bug` label to this PR? the `mergable` CI test fails due to that missing classification.

ax3l commented on setup.py:
```diff
@@ -36,8 +37,20 @@
     long_description = file.read()
 
+OMP_CONFIG = defaultdict(
+    lambda: ["-fopenmp"], {"unix": ["-fopenmp"], "msvc": ["/openmp"]}
```
unix: this is a bit optimistic but will work for GCC and Clang at least

ax3l replied:
furthermore: AppleClang does not support OpenMP by default.
If you want to unconditionally add an OpenMP prerequisite, you could install `brew install libomp` on CI (and need to document this new dependency on macOS).

Practically, one might want to implement a compiler check and keep OpenMP optional.

ax3l commented:
@cphyc if compile errors occur in the yt cythonization process then the overall build process does not always abort. I am no cython / distutils expert, but ideally this should be addressed independently of this PR, otherwise users get faulty installs because such errors are not aborting the build process and get lost between many build output messages and compiler warnings in the code base.

RevathiJambunathan commented:
@ax3l Thank you for this PR!
I tried installing yt from your branch and I am able to successfully import it.
```
(yt_env) rjambunathan@zilla:~/Documents/SOFTWARES/yt$ python
Python 3.7.9 (default, Aug 18 2020, 06:24:24) 
[GCC 5.4.0 20160609] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import yt
>>> yt.__version__
'4.0.dev0'
```

Xarthisius commented on setup.py:
```diff
@@ -36,8 +37,20 @@
+
+_COMPILER = get_default_compiler()
+
 if check_for_openmp():
```
`check_for_openmp()` makes some assumptions about openmp flags. As a result your code will only be called only if `-fopenmp` works, otherwise it will default to `[]`

ax3l replied:
yes, the assumption about flags (tested and applied) should be in one place, e.g. `check_for_openmp`. we currently duplicate the flags (and assumptions).

pep8speaks commented:
Hi there, @ax3l! Thanks for updating this PR. 



* In the file [`setupext.py`](https://github.com/yt-project/yt/blob/86f7617d79b307d8db250ee8036ebcdbb7fa6666/setupext.py):

> [Line 143:9](https://github.com/yt-project/yt/blob/86f7617d79b307d8db250ee8036ebcdbb7fa6666/setupext.py#L143): [E113](https://duckduckgo.com/?q=pep8%20E113) unexpected indentation

ax3l commented:
@cphyc I simplified and fixed the compiler logic changes you pushed on this PR.

If you don't mind, I would love to not add features in this PR. If you like to widen the support for compilers and OpenMP to more platforms, can we do so in a follow-up PR? Let's focus on the bugfix here.

ax3l commented on setup.py:
```diff
@@ -65,6 +70,7 @@
     "FIXED_INTERP": "yt/utilities/lib/fixed_interpolator.cpp",
     "ARTIO_SOURCE": glob.glob("yt/frontends/artio/artio_headers/*.c"),
     "CPP14_FLAG": CPP14_FLAG,
+    "CPP03_FLAG": CPP03_FLAG,
```
If possible, build all binaries with the same C++ standard. Especially between C++03 and C++11 were ABI breaks in the STL and with MSVC the math libs and STL will also cause problems.

cphyc replied:
I believe it is fine to replace the `CPP03_flag` with the `CPP14_flag` in the two files that use it. I added commit suggestions where relevant.
```suggestion
```

matthewturk commented:
Oh my goodness, thank you for taking such a close look at this.  C++ ABIs
and whatnot are things I had not even considered when I was looking at the
level of C++ support we needed.

(Although, maybe I'm also a bit oversensitive on this because I remember
the endless discussion of C++ ABI breakages back in the early 2000s... ;-)

On Fri, Oct 9, 2020 at 3:19 PM Axel Huebl <notifications@github.com> wrote:

> *@ax3l* commented on this pull request.
> ------------------------------
>
> In setup.py
> <https://github.com/yt-project/yt/pull/2939#discussion_r502652756>:
>
> > @@ -65,6 +70,7 @@
>      "FIXED_INTERP": "yt/utilities/lib/fixed_interpolator.cpp",
>      "ARTIO_SOURCE": glob.glob("yt/frontends/artio/artio_headers/*.c"),
>      "CPP14_FLAG": CPP14_FLAG,
> +    "CPP03_FLAG": CPP03_FLAG,
>
> If possible, build all binaries with the same C++ standard. Especially
> between C++03 and C++11 were ABI breaks in the STL and with MSVC the math
> libs and STL will also cause problems.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/yt-project/yt/pull/2939#pullrequestreview-505965298>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAVXO624F6MPWL7MEKRIRTSJ5V4HANCNFSM4SJICCGA>
> .
>

ax3l commented:
Did something change in `pykdtree` recently? It seems to also try to unconditionally build (as a dependency of yt) with `-fopenmp` flags, which fails on macOS.

ax3l commented:
On windows, it looks like we need to potentially remove some of the older work-arounds in `yt/utilities/lib/platform_dep.h`?

cphyc commented on yt/utilities/lib/cykdtree/kdtree.pyx:
```diff
@@ -3,7 +3,7 @@
 # distutils: depends = yt/utilities/lib/cykdtree/c_kdtree.hpp, yt/utilities/lib/cykdtree/c_utils.hpp
 # distutils: language = c++
-# distutils: extra_compile_args = -std=c++03
+# distutils: extra_compile_args = CPP03_FLAG
```
```suggestion
# distutils: extra_compile_args = CPP14_FLAG
```

cphyc commented on yt/utilities/lib/cykdtree/utils.pyx:
```diff
@@ -2,7 +2,7 @@
 # distutils: depends = yt/utilities/lib/cykdtree/c_utils.hpp
 # distutils: language = c++
-# distutils: extra_compile_args = -std=c++03
+# distutils: extra_compile_args = CPP03_FLAG
```
```suggestion
# distutils: extra_compile_args = CPP14_FLAG
```

cphyc commented on yt/utilities/lib/geometry_utils.pyx:
```diff
@@ -1,6 +1,6 @@
 # distutils: libraries = STD_LIBS
-# distutils: extra_compile_args = OMP_ARGS
-# distutils: extra_link_args = OMP_ARGS
+# distutils: extra_compile_args = CPP14_FLAG OMP_ARGS
```
I believe adding the CPP14_FLAG doesn't do anything unless you also specify the language, doesn't it?

```suggestion
# distutile: language = c++
# distutils: extra_compile_args = CPP14_FLAG OMP_ARGS
```

neutrinoceros replied:
careful there, you misspelled `distutils` in your suggestion :)

cphyc commented:
> @cphyc I simplified and fixed the compiler logic changes you pushed on this PR.
> 
> If you don't mind, I would love to not add features in this bugfix PR. If you like to widen the support for compilers and OpenMP to more platforms, can we do so in a follow-up PR? I would suggest to improve the compile tests performed in `check_for_openmp` by just testing more known flags of popular compilers. Let's focus on the bugfix here.

Sure! I initially did not wish to push to your PR, so feel free to take it in any direction you think is the best!

ax3l commented:
Hm, I wonder why macOS still tries to build with OpenMP and what is wrong with the `erf` on Windows?

cphyc commented:
> Hm, I wonder why macOS still tries to build with OpenMP and what is wrong with the `erf` on Windows?

According to https://github.com/storpipfugl/pykdtree/issues/51, this is an issue on pykdtree's side. One solution is to not use openmp when compiling on MacOSX. This can be done by setting `USE_OMP=0` in the environment variables.

I have tried that in https://github.com/cphyc/yt/commit/2ad8d1b066a07d1fc27fd0fe38fcee019cac2803, and it seems to do the trick!

cphyc commented:
Good news, with the HEAD of https://github.com/cphyc/yt/tree/fix-gcc5 (https://github.com/cphyc/yt/commit/af550d9b0098acbe4f69854d285759d9638e2866), the OSX job now builds! Note that it still doesn't work with Windows.

cphyc commented:
I've issues my fixes in another PR (#2943), which hopefully should contribute to make the tests pass on this one.

cphyc commented on yt/utilities/lib/field_interpolation_tables.pxd:
```diff
@@ -10,7 +13,8 @@ cimport cython
 from yt.utilities.lib.fp_utils cimport imax, fmax, imin, fmin, iclip, fclip, fabs
 from libc.stdlib cimport malloc
-from libc.math cimport isnormal
+cdef extern from "<cmath>" namespace "std":
+    bint isnormal(double x) nogil
```
Note: for some reason, Visual Studio doesn't happy about this. However, if we replace by
```suggestion
cdef extern from "<math.h>" namespace "std":
    bint isnormal(double x) nogil
```
it compiles.

cphyc reviewed (approved):
Provided the tests pass, I would be happy to merge this!

cphyc commented:
/isort

cphyc commented:
@neutrinoceros @Xarthisius @ax3l  I have pushed some more commits to make the Windows build build, do you think this is ready to go? I have added some code myself so I don't feel like pushing the merge button ;)

Xarthisius commented on setup.py:
```diff
@@ -36,20 +37,24 @@
+
+_COMPILER = get_default_compiler()
+
+omp_args, _ = check_for_openmp()
```
If we are distinguishing between compile and link flags inside `check_for_openmp()`, shouldn't we do the same thing during cythonization, i.e. use two separate aliases? Otherwise, why return two variables?

cphyc commented on yt/utilities/lib/cykdtree/utils.pyx:
```diff
@@ -2,7 +2,7 @@
 # distutils: depends = yt/utilities/lib/cykdtree/c_utils.hpp
 # distutils: language = c++
-# distutils: extra_compile_args = -std=c++03
+# distutils: extra_compile_args = CPP14_FLAG
```
It depends, this was changed assuming that one flag was better than two. In this approach, the default C++ language version is C++14. If we want to make minimal changes, then this should be replaced by `CPP03_FLAG`. What do you think is best?

munkm commented:
/isort

ax3l commented:
Thank you everyone for getting this PR done and merged.

@RevathiJambunathan just a last check, does the yt `master` branch compile again on your system with GCC 5?

munkm commented:
Thank *you* so much for the fix!! We really appreciate your contribution.
